arch/arm/core/swap.S

```diff
@@ -283,12 +294,18 @@ SECTION_FUNC(TEXT, __svc)
     * 0: context switch
     * 1: irq_offload (if configured)
     * 2: kernel panic or oops (software generated fatal exception)
+    * 3: System call
     * Planned implementation of system calls for memory protection will
     * expand this case.
     */
     ands r1, #0xff
     beq _context_switch
 
+#if CONFIG_USERSPACE
+    cmp r1, #3
+    beq _do_syscall
+#endif
+
```

Comment:
Would it be possible to check up front if the caller is unpriv and only allow it to use svc 3 in that case?  Otherwise, we're allowing userspace to invoke any of the other svc calls, and some of those might be trusting the caller.

Reply:
Sure, the mode is in the CONTROL reg, and I did this check and bombed out in a previous version of this if the code was privileged.  I am on the fence on this one

Reply:
My concern is different - I don't care about preventing a privileged caller from making an arbitrary svc call; I just want to prevent an unprivileged caller from calling anything other than svc 3 (and any other individual svcs that are explicitly whitelisted as being safe for use by userspace).

Reply:
Ah ok I see what you mean.  This would mean some sort of macro or interface where we define system call numbers and their access mode and then do a lookup to validate it in the actual svc call.

Reply:
I don't think that's what Stephen is saying.
What we want to prevent is user mode from invoking (for example) SVC #1. That would be a backdoor into the system.

Reply:
The whole point of the SVC is to run something as superuser.  That we use it today to do odd things like context switching or offloading while we are already privileged is kind or neither here nor there.  I don't know how you would enforce being able to service a SVC or not.  Sure we can add a lot of checks to make sure we are in the right 'state' when we make specific calls, but really,  I don't see what that gains us.  An SVC should do one specific thing and then return.

Reply:
Then I think we need to refactor the design.
Right now, we are using SVC for stuff which should never be allowable from user mode: irq_offload() and inducing a kernel panic. irq_offload runs an arbitrary function pointer in interrupt context. kernel panic hoses the entire system (unlike an oops which just kills a thread).
Is there no way to check what the interrupting context's privilege level was?

Reply:
Yes you know the callers privilege level.  My point is that the SVC has always been meant to be called from unprivileged contexts.  If we want to add this extra consideration, so be it.

Reply:
> My point is that the SVC has always been meant to be called from unprivileged contexts

For the irq_offload() and panic cases, we are using SVC because we want to generate a CPU exception from software, even though these should not be allowable from user mode.
Is there perhaps a better way to do it?
For sake of efficiency, we could check that the CPU is not in user mode in the logic for panic/offload cases and jump to the oops case if the test fails.

Reply:
And what do we return if we invoke a syscall from privileged context?  We shouldn't allow that either.

Reply:
> And what do we return if we invoke a syscall from privileged context? We shouldn't allow that either.

Do we care?
AFAIK on x86 it does work, although it's a terribly inefficient thing to do since you can just call the implementation directly.
I'm fine with leaving the scenario as "undefined behavior" and not doing checks, or only checking in an assertion. I can't conceive on how supervisor mode would be invoking system calls like this unless someone is doing something really weird, and we don't try to prevent malfeasance in supervisor mode anyway.